File: typescript/nucleotide-count/nucleotide-count.ts
export const nucleotideCounts = (s: string) => {
    const MAP: any = {'G': 0, 'A': 0, 'T': 0, 'C': 0}
    s.split('').forEach(c => {
        if (c in MAP) MAP[c]++; else throw new Error("Invalid nucleotide in strand");
    })

    return MAP;

};

File: dart/nucleotide-count/lib/nucleotide_count.dart
class InvalidNucleotideException implements Exception {}

class NucleotideCount {
  Map<String, int> count(String strand) {
    
    final result = {"A": 0, "C": 0, "G": 0, "T": 0};
    
    if (strand.split('').any((x) => !result.containsKey(x))) throw InvalidNucleotideException();
    
    return strand.split('').fold(result, (acc, n) { acc[n] = acc[n]! + 1; return acc; });
    
  }
}
